src/Tools/BuildValidator/CompilationDiff.cs

```diff
@@ -26,44 +37,163 @@ internal class CompilationDiff
             OriginalPath = originalPath;
         }
 
-        private CompilationDiff(
-            string originalPath,
-            Exception exception)
-        {
-            OriginalPath = originalPath;
-            Exception = exception;
-        }
-
         private CompilationDiff(ImmutableArray<Diagnostic> diagnostics, string originalPath)
         {
             Diagnostics = diagnostics;
             OriginalPath = originalPath;
         }
 
-        public static CompilationDiff Create(FileInfo assemblyFile, Compilation producedCompilation)
+        public static unsafe CompilationDiff Create(
+            FileInfo originalBinaryPath,
+            CompilationOptionsReader optionsReader,
+            Compilation producedCompilation,
+            IMethodSymbol? debugEntryPoint,
+            Options options)
         {
-            using var peStream = new MemoryStream();
+            using var rebuildPeStream = new MemoryStream();
+
+            using var win32ResourceStream = producedCompilation.CreateDefaultWin32Resources(
+                versionResource: true,
+                noManifest: producedCompilation.Options.OutputKind == OutputKind.DynamicallyLinkedLibrary,
+                manifestContents: null,
+                iconInIcoFormat: null);
+
+            var sourceLink = optionsReader.GetSourceLinkUTF8();
+            var emitResult = producedCompilation.Emit(
+                peStream: rebuildPeStream,
+                pdbStream: null,
+                xmlDocumentationStream: null,
+                win32Resources: win32ResourceStream,
+                manifestResources: optionsReader.GetManifestResources(),
+                options: new EmitOptions(
+                    debugInformationFormat: DebugInformationFormat.Embedded, highEntropyVirtualAddressSpace: true),
+                debugEntryPoint: debugEntryPoint,
+                metadataPEStream: null,
+                pdbOptionsBlobReader: optionsReader.GetMetadataCompilationOptionsBlobReader(),
```

Comment:
This still seems odd to me. We're essentially saying that regardless of what you actually use to compile, we're going to copy some random metadata from here and put it in the pdb.

Reply:
The options blob primarily exists to tell the provenance host how best to setup the environment for validation. If we force the compiler to re-compute the blob on verification then we are essentially mandating that the validation environment **exactly** match the original build one. That means the more specific we make information the more constraints we put on the host. 

I prefer this approach because we're giving discretion ability to the host. We're letting them determine how much effort they want to put into reproducing the environment. If that's not enough and it causes validation to fail then that's on the host for making that choice.

Reply:
Another way to think about it is the following. 

This will never cause a case which shouldn't validate to incorrectly validate (does not introduce false positive). Doing that would be detrimental to the product. 

This will cause causes which only fail to verify because the options blob was different to pass (removes false negatives). If literally everything but the options blob is the same then provenance has been verified. The only difference is the options blob is telling us several of the extra options, and versions, were different. That difference clearly didn't factor into the production of the PE or PDB.

Reply:
The options blob being different could be a good diagnostic tool if we do have a failure, or at least that was my thinking. If we're validating the binary without the options blob can we explicitly do that in the tool rather than introducing different ways to emit? 

To be clear, I'm not opposed to _what_ we are validating, just the introduction of an options blob as part of the Emit. I'm trying to think of this as a developer without knowledge of this feature tool, and I would be very confused by it.

Reply:
> The options blob being different could be a good diagnostic tool if we do have a failure, or at least that was my thinking.

I agree but these are not mutually exclusive options. Consider that the compiler knows how to generate the options blob for the current environment. To me the simplest approach to validation is:

Use original options blob to run validation. If that succeeds then great we're done. If this fails then we can have the compiler produce the options blob for the current environment, diff that and give that diagnostic to the customer. This is fairly easy to do, it's a small refactor, and then diff the options. 

> I'm trying to think of this as a developer without knowledge of this feature tool, and I would be very confused by it.

I don't think this impacts the end developer experience. I envision a diff of the options blob as part of the experience. It can even be done in the success case (imagine a `/pedantic` option where we always print out the diff of the environments